libevmasm/SemanticInformation.h

```diff
@@ -56,6 +56,17 @@ struct SemanticInformation
 	/// without altering the semantics. This means it cannot depend on storage or memory,
 	/// cannot have any side-effects, but it can depend on a call-constant state of the blockchain.
 	static bool movable(Instruction _instruction);
+	/// @returns true if the instruction can be moved or copied (together with its arguments) across
+	/// instructions and control flow branches that do not modify the blockchain state.
+	static bool movableIfStateInvariant(Instruction _instruction);
+	/// @returns true if the instruction can be moved or copied (together with its arguments) across
+	/// instructions and control flow branches that do not modify storage or the blockchain state.
+	static bool movableIfStorageInvariant(Instruction _instruction);
+	/// @return true if the instruction can be moved or copied (together with its arguments) across
+	/// instructions and control flow branches that do not modify memory or contain the msize
+	/// instruction.
+	static bool movableIfMemoryInvariant(Instruction _instruction);
```

Comment:
I wonder if we the name should also contain MSize. `movableIfMemoryInvariantAndNoMSize` seems too much?